bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -276,21 +284,23 @@ private HeaderCacheElement getCachedHeader() throws IOException {
    */
   @VisibleForTesting
   boolean doRefresh() {
-    boolean requiresRefresh = false;
     synchronized (isRefreshing) {
       if (!isRefreshing.get() && getCacheState(this.headerCache.get()) != CacheState.Good) {
         isRefreshing.set(true);
-        requiresRefresh = true;
+      } else {
+        return false;
       }
     }
-    if (!requiresRefresh) {
-      return false;
-    }
-    HeaderCacheElement cacheElement = refreshCredentialsWithRetry();
-    synchronized (isRefreshing) {
-      headerCache.set(cacheElement);
-      isRefreshing.set(false);
-      isRefreshing.notifyAll();
+    try {
+      HeaderCacheElement cacheElement = refreshCredentialsWithRetry();
+      synchronized (isRefreshing) {
```

Comment:
Sorry, yeah I understand the intent, what I mean is that I don't think the headerCache call specifically needs to be inside the block. You've already done the refreshCredentialsWithRetry outside the lock so this is just populating an atomic reference. 

A different way of asking this is actually, shouldn't refreshCredentialsWithRetry be *inside* the synchronized block? Otherwise you're still kicking off a bunch of requests?